### PR TITLE
refactor(SPV-1079): move merkleroots to spv-wallet service

### PR DESCRIPTION
**Pull request by dzolt-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

**Review by Nazarii-4chain — APPROVED** [review]
LGTM

File: engine/chain/internal/bhs/get_merkleroots_from_bhs_test.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/bitcoin-sv/spv-wallet/engine/chain"
+	"github.com/bitcoin-sv/spv-wallet/engine/chain/errors"
+	"github.com/bitcoin-sv/spv-wallet/engine/chain/models"
+	"github.com/bitcoin-sv/spv-wallet/engine/tester"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetMerkleRootsFromBHSSuccess(t *testing.T) {
+	t.Run("Get MerkleRoots success", func(t *testing.T) {
```
Comment (dorzepowski) [inline engine/chain/internal/bhs/get_merkleroots_from_bhs_test.go:18]: Because it is used only for kinda proxy or reading, I would prefer to add `integration`/`endpoint` tests instead unit tests. 

If you want to make this change in a separate PR. So it wouldn't block this PR.

File: engine/chain/errors/bhs_error_response.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+	chainmodels "github.com/bitcoin-sv/spv-wallet/engine/chain/models"
+	"github.com/bitcoin-sv/spv-wallet/engine/spverrors"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/go-resty/resty/v2"
+)
+
+// MapBHSErrorResponseToSpverror is a method that will check what kind of response came back from
+// Block Header Service and map it to spverror and set it to context
+func MapBHSErrorResponseToSpverror(res *resty.Response) models.SPVError {
```
Comment (dorzepowski) [inline engine/chain/errors/bhs_error_response.go:16]: Why this is public, does it need to be in package errors? for me it's rather a part of the logic of the service

File: engine/chain/models/bhs_error.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package chainmodels
+
+// BHSError is an error response that is returned from Block Header Service (BHS)
+type BHSError struct {
```
Comment (dorzepowski) [inline engine/chain/models/bhs_error.go:4]: is it expose anywhere by a service, maybe it could be private and in the bhs package?

File: actions/merkleroots/get.go [hunk 4]
```diff
@@ -6,18 +6,30 @@ import (
...
+// @Param		lastEvaluatedKey query string false "last processed merkleroot in client's database"
+// @Success		200 {object} models.MerkleRootsBHSResponse "Paged response with Merkle Roots array in content "
+// @Failure		400	"Bad request - batchSize must be 0 or a positive integer"
+// @Failure		404	"Not found - No block with provided merkleroot was found"
+// @Failure		409	"Conflict - Provided merkleroot is not part of the longest chain"
+// @Failure 	500	"Internal Server Error - cannot create Block Header Service url. Please check your configuration"
+// @Failure 	500	"Internal Server Error - Block Header Service cannot be requested"
+// @Failure 	500	"Internal Server Error - Error while fetching Merkle Roots"
+// @Router  /api/v1/merkleroots [get]
+// @Security	x-auth-xpub
```
Comment (dorzepowski) [inline actions/merkleroots/get.go:27]: you need to generate swagger docs

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/chain/errors/bhs_error_response.go (hunk 2 above)
Comment (dzolt-4chain) [reply engine/chain/errors/bhs_error_response.go:16]: I made it public so it can be imported in bhs service, and put it in error package, because it seemed like a good place to handle error responses.  It is indeed only used in a service so I will move it

File: engine/chain/models/bhs_error.go (hunk 3 above)
Comment (dzolt-4chain) [reply engine/chain/models/bhs_error.go]: moved

File: engine/chain/errors/bhs_error_response.go (hunk 2 above)
Comment (dzolt-4chain) [reply engine/chain/errors/bhs_error_response.go]: moved

File: actions/merkleroots/get.go (hunk 4 above)
Comment (dzolt-4chain) [reply actions/merkleroots/get.go:27]: done

**Review by dorzepowski — APPROVED** [review]

File: engine/chain/interface.go [hunk 5]
```diff
@@ -15,6 +17,7 @@ type ARCService interface {
 
 // BHSService for querying BHS server.
 type BHSService interface {
+	GetMerkleRootsFromBHS(ctx context.Context, query url.Values) (*models.MerkleRootsBHSResponse, error)
```
Comment (chris-4chain) [inline engine/chain/interface.go:20]: I would not add the `FromBHS` suffix as other methods (within BHSSErvice or ARCService) don't have such suffix. 
So I vote for `GetMerkleRoots`.

File: engine/chain/internal/bhs/bhs_error.go [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package bhs
+
+// bhsError is an error response that is returned from Block Header Service (BHS)
+type bhsError struct {
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/bhs_error.go:4]: Private (unexported) structs don't require description/comments. 
If you really want to keep it, please make it more descriptive and less obvious.

File: engine/chain/internal/bhs/bhs_error_response.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+// mapBHSErrorResponseToSpverror is a method that will check what kind of response came back from
+// Block Header Service and map it to spverror and set it to context
+func mapBHSErrorResponseToSpverror(res *resty.Response) models.SPVError {
+	var responseErr bhsError
+
+	if err := json.Unmarshal(res.Body(), &responseErr); err != nil {
+
+		// TODO: remove after SPV-1106 is done
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/bhs_error_response.go:21]: Please add appropriate task to jira to not forget about this `TODO`.
- or point out in `SPV-1106` description to do so 😄

File: engine/chain/internal/bhs/bhs_error_response.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+	switch responseErr.Code {
+	case "ErrInvalidBatchSize":
+		return chainerrors.ErrInvalidBatchSize
+	case "ErrMerkleRootNotFound":
+		return chainerrors.ErrMerkleRootNotFound
+	case "ErrMerkleRootNotInLC":
+		return chainerrors.ErrMerkleRootNotInLongestChain
+	case "error-unauthorized":
+		return chainerrors.ErrBHSUnauthorized
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/bhs_error_response.go:47]: Maybe strings "enum" would be good here?

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		req.SetAuthToken(s.bhsCfg.AuthToken)
+	} else {
+		s.logger.Warn().Msg("warning creating Block Headers Service url - auth token is not set. Some requests might not work")
+	}
+
+	var response models.MerkleRootsBHSResponse
+	res, err := req.
+		SetResult(&response).
+		SetQueryParamsFromValues(query).
+		Get(bhsURL.String())
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/get_merkleroots_from_bhs.go:37]: In the `chain` package I've used the convention in which the `Get/Post` method call is not mexed up with the `req` configuration (e.g. SetResult`).

After adjustments it would look like this:
```golang
        var response models.MerkleRootsBHSResponse
	req := s.httpClient.R().
		SetContext(ctx).
		SetHeader("Content-Type", "application/json").
		EnableTrace().
		SetResult(&response).
		SetQueryParamsFromValues(query)

	if s.bhsCfg.AuthToken != "" {
		req.SetAuthToken(s.bhsCfg.AuthToken)
	} else {
		s.logger.Warn().Msg("warning creating Block Headers Service url - auth token is not set. Some requests might not work")
	}

	
	res, err := req.Get(bhsURL.String())
```

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+	"syscall"
+
+	"github.com/bitcoin-sv/spv-wallet/engine/chain/errors"
+	"github.com/bitcoin-sv/spv-wallet/engine/spverrors"
+	"github.com/bitcoin-sv/spv-wallet/models"
+)
+
+// GetMerkleRootsFromBHS returns Merkle Roots from Block Header Service
+func (s *Service) GetMerkleRootsFromBHS(ctx context.Context, query url.Values) (*models.MerkleRootsBHSResponse, error) {
+	bhsURL, err := s.createBHSURL("/chain/merkleroot")
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/get_merkleroots_from_bhs.go:17]: You could adjust also `makeVerifyMerkleRootsRequest` to use the `createBHSURL` helper method.

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	req := s.httpClient.R().
+		SetContext(ctx).
+		SetHeader("Content-Type", "application/json").
+		EnableTrace()
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/get_merkleroots_from_bhs.go:25]: I'm not convinced if we need the `EnableTrace` here. I didn't apply it for other `resty`-requsts in the `chain` package. 
hmmm 🤔 

Maybe we could rather use [client.EnableTrace](https://pkg.go.dev/github.com/go-resty/resty/v2#Client.EnableTrace) for the whole resty client - and decide whether or not to use it on the level where the client is created ?

I'm wondering how this influence to performance. Perhaps we should enable trace only when `logger level is set to debug` 🤔

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+	var response models.MerkleRootsBHSResponse
+	res, err := req.
+		SetResult(&response).
+		SetQueryParamsFromValues(query).
+		Get(bhsURL.String())
+	if err != nil {
+		if errors.Is(err, syscall.ECONNREFUSED) {
+			return nil, chainerrors.ErrBHSUnreachable.Wrap(err)
+		}
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/get_merkleroots_from_bhs.go:41]: I did similar thing in my implementations. But slightly different.
I think, we should pick one. (I don't know, for now, which one is better  😄 )

My checking of error:
```golang
if err != nil {
		var e net.Error
		if errors.As(err, &e) {
			return nil, chainerrors.ErrBHSUnreachable.Wrap(err)
		}
		return nil, spverrors.ErrInternal.Wrap(err)
	}
```

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+	return &response, nil
+}
+
+// createBHSURL parses Block Header Url from configuration and constructs a valid
+// endpoint with provided endpointPath variable
+func (s *Service) createBHSURL(endpointPath string) (*url.URL, error) {
+	if s.bhsCfg.URL == "" {
+		s.logger.Error().Msgf("create Block Header Service URL - url not configured")
+	}
```
Comment (chris-4chain) [inline engine/chain/internal/bhs/get_merkleroots_from_bhs.go:56]: I didn't check the URL at this level. And I'm not convinced we have to do so 🤔 
Besides, we do the verification at `config` level:
https://github.com/bitcoin-sv/spv-wallet/blob/ce45722045af9f82c57d4d8935492b52d4dd0e17/config/validate_arc.go#L21-L23

**Review by chris-4chain — CHANGES_REQUESTED** [review]

File: engine/chain/internal/bhs/bhs_error_response.go (hunk 7 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/bhs_error_response.go]: after discussion, removed the comment and left this handle error here, It might be useful when server returns 404 not foundd which is a string by default

File: engine/chain/internal/bhs/bhs_error_response.go (hunk 8 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/bhs_error_response.go:46]: These are only cases that will occur from bhs, I'd honestly leave it as is but if you think there are any benefits of using string enums here let me know and i might change it.

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go (hunk 10 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/get_merkleroots_from_bhs.go]: done

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go (hunk 11 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/get_merkleroots_from_bhs.go]: Yeah, you're right it should rather be set for the client and not per request. I will remove it from here and I suggest we discuss it if we even want it set or in a particular scenario as you mentioned

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go (hunk 9 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/get_merkleroots_from_bhs.go]: done

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go (hunk 12 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/get_merkleroots_from_bhs.go]: Hmm, my approach will handle more narrow cases like when server is unresponsive, while yours will also handle cases like timeouts and temporary network failures, so i incline more towards your approach. I will adjust mine

File: engine/chain/internal/bhs/get_merkleroots_from_bhs.go (hunk 13 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/get_merkleroots_from_bhs.go]: mm yeah, I copied what I created in a previous version but since we are checking it at the config level there is no need to do it twice, removed

File: engine/chain/internal/bhs/get_merkleroots_from_bhs_test.go (hunk 1 above)
Comment (dzolt-4chain) [reply engine/chain/internal/bhs/get_merkleroots_from_bhs_test.go]: okay, lets make it in a separate pr

File: engine/chain/internal/bhs/bhs_error_response.go (hunk 8 above)
Comment (chris-4chain) [reply engine/chain/internal/bhs/bhs_error_response.go:46]: It's ok 👍

**Review by chris-4chain — APPROVED** [review]
